src/providers/aws.rs

```diff
@@ -969,130 +970,48 @@ impl RegionLauncher {
-                            None
-                        }
+                    .map(|(request_id, state, instance_id)| {
+                        assert_eq!(state, "active");
+                        let instance_id = instance_id.unwrap();
+                        let setup = self.spot_requests.get(&request_id).cloned().unwrap();
+                        (instance_id, setup)
                     })
                     .collect();
                 break;
-            } else {
-                tokio::time::delay_for(time::Duration::from_secs(1)).await;
             }
 
             if let Some(wait_limit) = max_wait {
                 if start.elapsed() <= wait_limit {
                     continue;
                 }
-
-                tracing::warn!("wait time exceeded -- cancelling run");
-                let mut cancel = rusoto_ec2::CancelSpotInstanceRequestsRequest::default();
-                cancel.spot_instance_request_ids = req
-                    .spot_instance_request_ids
-                    .clone()
-                    .expect("we set this to Some above");
-                client
-                    .cancel_spot_instance_requests(cancel)
-                    .await
-                    .wrap_err("failed to cancel spot instances")?;
-
-                tracing::trace!("spot instances cancelled -- gathering remaining instances");
-                // wait for a little while for the cancelled spot requests to settle
-                // and any that were *just* made active to be associated with their instances
-                tokio::time::delay_for(time::Duration::from_secs(1)).await;
-
-                let sirs = client
-                    .describe_spot_instance_requests(req)
-                    .await
-                    .wrap_err("failed to inspect canceled spot requests")?
-                    .spot_instance_requests
-                    .unwrap_or_else(Vec::new);
-                for sir in sirs {
-                    let req_id = sir.spot_instance_request_id.unwrap();
-                    match sir.instance_id {
-                        Some(instance_id) => {
-                            tracing::trace!(
-                                %req_id,
-                                iid = %instance_id,
-                                "spot request cancelled"
-                            );
-                        }
-                        _ => {
-                            tracing::error!(
-                                %req_id,
-                                "spot request failed: {:?}", &sir.status,
-                            );
-                        }
-                    }
-                }
+                self.cancel_spot_instance_requests(&request_ids).await?;
                 eyre::bail!("wait limit reached");
             }
+
+            // let's not hammer the API
+            tokio::time::delay_for(time::Duration::from_secs(1)).await;
```

Comment:
This will now not be hit if `wait_limit.is_some()` and `start.elapsed() <= wait_limit`, since that `continue`s early.